arch/x86/core/userspace.S

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2017 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel_structs.h>
+#include <arch/x86/asm.h>
+#include <offsets_short.h>
+
+GTEXT(_x86_syscall_invoke)
+GTEXT(_x86_syscall_entry_stub)
+GTEXT(_x86_syscall_entry)
+GTEXT(_x86_userspace_enter)
+
+/* Landing site for syscall SW IRQ. Marshal arguments and call C function for
+ * further processing.
+ */
+SECTION_FUNC(TEXT, _x86_syscall_entry_stub)
+	push	%esi		/* arg5 */
+	push	%edi		/* arg4 */
+	push	%ebx		/* arg3 */
+#ifndef CONFIG_X86_IAMCU
+	push	%ecx		/* arg2 */
+	push	%edx		/* arg1	*/
+	push	%eax		/* Call id */
+#endif
+
+	call	_x86_syscall_entry
+
+	/* EAX now contains return value. Pop or xor everything else to prevent
+	 * information leak from kernel mode.
+	 */
+#ifndef CONFIG_X86_IAMCU
+	add	$4, %esp	/* old EAX value */
+	pop	%edx
+	pop	%ecx
+#else
+	xor	%edx, %edx
+	xor	%ecx, %ecx
+#endif
+	pop	%ebx
+	pop	%edi
+	pop	%esi
+
+	iret
+
+
+/* FUNC_NORETURN void _x86_userspace_enter(_thread_entry_t user_entry,
+ *					   void *p1, void *p2, void *p3,
+ *					   u32_t stack_end,
+ *					   u32_t stack_start)
+ *
+ * A one-way trip to userspace.
+ */
+SECTION_FUNC(TEXT, _x86_userspace_enter)
+	add	$0x4, %esp /* Discard return address on stack */
+
+	/* Fetch parameters on the stack */
+#ifndef CONFIG_X86_IAMCU
+	pop	%eax	/* user_entry */
+	pop	%edx	/* p1 */
+	pop	%ecx	/* p2 */
+#endif
+	pop	%esi	/* p3 */
+	pop	%ebx	/* stack_end (high address) */
+	pop	%edi	/* stack_start (low address) */
+
+	/* Move to the kernel stack for this thread, so we can erase the
+	 * user stack. The kernel stack is the page immediately before
+	 * the user stack.
+	 *
+	 * For security reasons, we must erase the entire user stack.
+	 * We don't know what previous contexts it was used and do not
+	 * want to leak any information.
+	 */
+	mov	%edi, %esp
+
+	/* Stash some registers we are going to need to erase the user
+	 * stack.
+	 */
+	push	%ecx
+	push	%edi
+	push	%eax
+
+	/* Compute size of user stack and put in ECX */
+	mov	%ebx, %ecx
+	sub	%edi, %ecx
+
+#ifdef CONFIG_INIT_STACKS
+	mov	$0xAAAAAAAA, %eax
```

Comment:
Another good reason not to duplicate the stack clear from C is that you need duplicate handling for all the somewhat obscure features.

Reply:
you've confused me here.
there is no C code specifically for clearing stacks.